wire-desktop: refactor to add Darwin support

toonn opened the pull request:
###### Motivation for this change
The derivation didn't have Darwin support but the software does.
I refactored the package putting all the linux/darwin specific stuff in their own attrsets and then that's  merged into the common bits like `pname` and `meta`.

I updated the `meta` while I was at it. The license in the repo is actually GPLv3 *or later* but I'm not sure whether that's accurate because terms and conditions apply. 

You can find them in the [README for the project](https://github.com/wireapp/wire-desktop/blob/staging/README.md), reproduced here for convenience:

  > For licensing information, see the attached LICENSE file and the list of third-party licenses at wire.com/legal/licenses/.

  > If you compile the open source software that we make available from time to time to develop your own mobile, desktop or web application, and cause that application to connect to our servers for any purposes, we refer to that resulting application as an “Open Source App”. All Open Source Apps are subject to, and may only be used and/or commercialized in accordance with, the Terms of Use applicable to the Wire Application, which can be found at https://wire.com/legal/#terms. Additionally, if you choose to build an Open Source App, certain restrictions apply, as follows:

  > a. You agree not to change the way the Open Source App connects and interacts with our servers;
    b. You agree not to weaken any of the security features of the Open Source App;
    c. You agree not to use our servers to store data for purposes other than the intended and original functionality of the Open Source App;
    d. You acknowledge that you are solely responsible for any and all updates to your Open Source App.

  > For clarity, if you compile the open source software that we make available from time to time to develop your own mobile, desktop or web application, and do not cause that application to connect to our servers for any purposes, then that application will not be deemed an Open Source App and the foregoing will not apply to that application.

  > No license is granted to the Wire trademark and its associated logos, all of which will continue to be owned exclusively by Wire Swiss GmbH. Any use of the Wire trademark and/or its associated logos is expressly prohibited without the express prior written consent of Wire Swiss GmbH.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) **Tested the executable inside the `.app`.**
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) **Tried to but the original derivation wouldn't build on darwin. The result is about 150MB though.**
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

armijnhemel commented:
adding @silverhook for legal review, as I am not a lawyer. I don't think that the restrictions that Wire has added would be considered permissable "additional terms" as per GPL3, section 7 but instead be a "further restriction".

silverhook commented:
I didn’t dive deep, but at a quick glance, I’d agree with @armijnhemel that this would go into the “further restrictions” territory. But also that these “further restrictions” fall away, if you modify the client in a way to not connect to Wire’s own servers.

It seems like a very mixed bag to me.

arianvp commented:
I work at Wire and have poked someone at the company who might be able to help with this question.

worldofpeace commented:
@toonn I've made a patch to fix some style issues and to make it easier to update manually
* https://gist.github.com/worldofpeace/d02bcda1ab44bb37e438335a46579232

toonn commented:
@worldofpeace, I applied your patch, I'd originally grouped the `makeDesktopItem` with the other linux-specific dependencies but I don't feel strongly about that. Keeping the versions together does look like a definite improvement.

raphaelrobert commented:
@toonn The client source code is only governed by GPLv3. The terms you quote apply to the usage of Wire servers, not the client server.

armijnhemel commented:
> @toonn The client source code is only governed by GPLv3. The terms you quote apply to the usage of Wire servers, not the client server.

Although I agree that these terms only seem to be when the Wire servers are used (and not when your own server is used), they *do* read on the client code and are, I think, a "further restriction" according to the license.

raphaelrobert commented:
I should have clarified: I work for Wire. The terms of use only govern Wire accounts (which is a prerequisite for using the Wire servers).
The source code is not affected by them.

worldofpeace commented on pkgs/applications/networking/instant-messengers/wire-desktop/default.nix:
```diff
@@ -1,103 +1,130 @@
-      --set-rpath "${rpath}:$out/opt/Wire" \
-      "$out/opt/Wire/wire-desktop"
+    unpackPhase = ''
+      /usr/bin/xar -xf $src
```
Why is this an absolute path?

toonn replied:
I'll delve into this more tomorrow. But when I drop the `/usr/bin` (and add xar as an argument and buildInput) nix-build seems to hang after the download. No idea what's going on there.

toonn replied:
Not sure what was going on yesterday, today it just works. Thanks for spotting the impurity : )

armijnhemel commented:
> I should have clarified: I work for Wire. The terms of use only govern Wire accounts (which is a prerequisite for using the Wire servers).
> The source code is not affected by them.

But it is not "terms of use"! The README actually calls them "restrictions" :-)

"Additionally, if you choose to build an Open Source App, certain restrictions apply"

I absolutely understand *why* you would want these restrictions as rogue code could interfere with your operations, but the way it is currently worded it reads on the client code and thus could be seen as a "further restriction" according to the license.

Personally I would move these restrictions to general terms and conditions and just terminate people's accounts if they don't comply with those rules. GPL3, section 6 would already allow you to do this:

"Access to a network may be denied when the modification itself materially and adversely affects the operation of the network or violates the rules and protocols for communication across the network."

It might be worthwhile to talk to an experienced open source lawyer. I would be more than happy to give some suggestions for various jurisdictions (Germany, Switzerland, Italy, Spain, UK, etc.)

grahamc commented on pkgs/applications/networking/instant-messengers/wire-desktop/default.nix:
```diff
@@ -1,103 +1,130 @@
+    platforms = [ "x86_64-darwin" "x86_64-linux" ];
   };
-}
+} // (if stdenv.isDarwin then darwin else linux))
```
This is very fishy.

toonn replied:
I refactored this. Should I submit an issue for the electron package? That's where I got the construction.

toonn committed:
wire-desktop: expand meta

Added a `longDescription` and `downloadPage`. Also added myself to
`maintainers`.

I fixed up the `license` because it's actually GPLv3 *or later*.
Note that terms and conditions apply though I don't feel like they
violate the GPL-ness of the code.

toonn committed:
wire-desktop: Add darwin support

Wire Desktop is available for linux, mac os and windows. I figured
adding darwin support would be cromulent. Note that the versions don't
align 100%, this is how it's released upstream.

I refactored the derivation to seperate all the linux-specific parts. I
also sorted the dependencies and grouped them.

The changes were based on the derivation for electron. I changed the
construction from calling `mkDerivation` on a conditional merger of two
sets by moving the `mkDerivation` calls into the conditional and up to
the local bindings for `linux` and `darwin`. This required moving
`pname` and `meta` up to local bindings.

silverhook commented:
> I should have clarified: I work for Wire. The terms of use only govern Wire accounts (which is a prerequisite for using the Wire servers).
> The source code is not affected by them.

@raphaelrobert, thanks for the clarification. But if that is case, please make sure to clarify it in the license itself and official documentation, instead of an issue buried on a source code forge.

raphaelrobert commented:
@silverhook I'm not sure what you mean exactly. The README file points to the license file directly for all source code licensing questions:

> For licensing information, see the attached LICENSE file and the list of third-party licenses at wire.com/legal/licenses/.

silverhook commented:
@raphaelrobert, the relevant part of the https://github.com/wireapp/wire-desktop/blob/staging/README.md reads as follows (emphasis mine):

> For licensing information, see the attached LICENSE file and the list of third-party licenses at wire.com/legal/licenses/.
> 
> If you compile the open source software that we make available from time to time to develop your own mobile, desktop or web application, and cause that application to connect to our servers for any purposes, we refer to that resulting application as an “Open Source App”. **All Open Source Apps are subject to, and may only be used and/or commercialized in accordance with, the Terms of Use applicable to the Wire Application, which can be found at https://wire.com/legal/#terms. Additionally, if you choose to build an Open Source App, certain restrictions apply, as follows:**
> 
> a. **You agree not to change the way the Open Source App connects and interacts with our servers; b. You agree not to weaken any of the security features of the Open Source App; c. You agree not to use our servers to store data for purposes other than the intended and original functionality of the Open Source App; d. You acknowledge that you are solely responsible for any and all updates to your Open Source App.**
> 
> For clarity, if you compile the open source software that we make available from time to time to develop your own mobile, desktop or web application, and do not cause that application to connect to our servers for any purposes, then that application will not be deemed an Open Source App and the foregoing will not apply to that application.

The parts emphasised in bold, seem to be connected to the license of the code, not the Terms of Service of the official Wire server(s). I agree that you clarification makes sense, but it is not (unambiguously) reflected in the `README.md`. The way it reads now is that it restricts the use of the code and splits the market into two, giving a different license to each part of the market.

It would be good to re-write the wording in order to clarify the distinction between the license of the code (GPL-3.0) and the terms of service that apply when a client connects to the official Wire servers.

raphaelrobert commented:
Thanks for the feedback @silverhook! I guess we could make it more clear that the README is just an extract of the Terms of Use.

armijnhemel commented:
> Thanks for the feedback @silverhook! I guess we could make it more clear that the README is just an extract of the Terms of Use.

Yes, this would, I think, suffice (as I already commented). It basically comes down to "if you mess with our operations, we will terminate your account" and that's perfectly fine. For the rest: what @silverhook said :-)

toonn commented:
Ping @worldofpeace.

worldofpeace commented:
@grahamc at https://github.com/NixOS/nixpkgs/pull/64931#discussion_r304112617 you said "this is very fishy".

I'm not sure what you meant by that and what to change, and @toonn changed their expression significantly https://github.com/NixOS/nixpkgs/pull/64931/commits/f5beda50ca643cb4a26321d513772186410cfec5.

toonn commented:
@worldofpeace, I made the changes in response to @grahamc's complaints. They didn't like the `common // if darwin then darwin else linux)`, preferring inheriting the common attributes. (aside: I personally don't like the `inherit` mechanism because `with` can bring things in scope implicitly so you can't always tell where inheritted attributes are coming from.)

@grahamc also didn't like that all this derivation does is fetch the `.app` for darwin, preferring homebrew for this. But the linux derivation essentially does the same and I wrote this because I want to move *away* from brew. (Note there was some support for having this in nixpkgs despite "just fetching the `.app`.")

worldofpeace commented:
Did you discuss in private? Because he only left that comment here.

At any rate I'm sure the changes are fine, going to give it the last look over since it has been a week here for me.

toonn commented:
We discussed in #nixos-dev or #nixos-chat, not quite sure. I could share the relevant parts of my logs if necessary. Note that these changes were made almost a week ago. I didn't ping until now because I figured people are busy and this isn't high priority anyway.

worldofpeace commented:
Thanks for clearing it up @toonn, it just confused me as to where the inspiration for these changes for coming from.

@GrahamcOfBorg build wire-desktop

worldofpeace commented on pkgs/applications/networking/instant-messengers/wire-desktop/default.nix:
```diff
@@ -1,103 +1,132 @@
+    downloadPage = https://wire.com/download/;
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers; [ toonn worldofpeace ];
+    platforms = [ "x86_64-darwin" "x86_64-linux" ];
```
We should set this independently in `darwin` and `linux`.

toonn replied:
Done, considered this metadata for the entire derivation, which is indeed compatible with both platforms now.

toonn replied:
May have gotten messy with my commit message, do I need to stick to the `package: reason` convention for all of them?

worldofpeace replied:
It should be fine since it's one package, squashing should be easy.

worldofpeace commented on pkgs/applications/networking/instant-messengers/wire-desktop/default.nix:
```diff
@@ -1,103 +1,135 @@
+  linux = stdenv.mkDerivation rec {
+    inherit pname version meta;
+
+    platforms = [ "x86_64-linux" ];
```
```suggestion
    meta.platforms = [ "x86_64-linux" ];
```

same for the other

toonn replied:
Can and should we use the same trick as for `version` and `sha256`? Or does that depend on `meta.platforms` somehow?

worldofpeace replied:
Well for those, it's just done for ease of manual maintainability. I don't expect to ever have to change this.

worldofpeace commented:
@GrahamcOfBorg build wire-desktop

worldofpeace commented:
@toonn Can you squash your commits?
I'm not sure where to go with the history so I'll leave this up to you.

toonn commented:
@worldofpeace, should we worry about the ofborg failure?

worldofpeace commented:
Oh, I just now noticed that.

FRidh commented:
I undid this change when merging staging-next into staging because of a merge conflict. Please open a new PR for staging to reapply the change.

FRidh commented:
Nevermind, it is in. However, there was a problem with evaluation, see 15564fbb4d9e349f75edc1d4791bd7b5f4865973.

toonn commented:
@FRidh, does this require further action? Would my original approach work better maybe? Then `meta.platforms` can be evaluated without running into this (if nix is lazy enough).
